MassStorageDriver.cpp

```diff
@@ -0,0 +1,596 @@
+	print_hexbytes(report, 8);
+	msControlCompleted = true;
+
+}
+
+void msController::callbackIn(const Transfer_t *transfer)
+{
+	println("msController CallbackIn (static)");
+	if (transfer->driver) {
+		print("transfer->qtd.token = ");
+		println(transfer->qtd.token & 255);
+		((msController *)(transfer->driver))->new_dataIn(transfer);
+	}
+}
+
+void msController::callbackOut(const Transfer_t *transfer)
+{
+	println("msController CallbackOut (static)");
+	if (transfer->driver) {
+		print("transfer->qtd.token = ");
+		println(transfer->qtd.token & 255);
+		((msController *)(transfer->driver))->new_dataOut(transfer);
+	}
+}
+
+void msController::new_dataOut(const Transfer_t *transfer)
+{
+	uint32_t len = transfer->length - ((transfer->qtd.token >> 16) & 0x7FFF);
+	println("msController dataOut (static)", len, DEC);
+	print_hexbytes((uint8_t*)transfer->buffer, (len < 32)? len : 32 );
+	msOutCompleted = true; // Last out transaction is completed.
+}
+
+void msController::new_dataIn(const Transfer_t *transfer)
+{
+	uint32_t len = transfer->length - ((transfer->qtd.token >> 16) & 0x7FFF);
+	println("msController dataIn (static): ", len, DEC);
+	print_hexbytes((uint8_t*)transfer->buffer, (len < 32)? len : 32 );
+	msInCompleted = true; // Last in transaction is completed.
+}
+
+// Initialize Mass Storage Device
+uint8_t msController::mscInit(void) {
+#ifdef DBGprint
+	Serial.printf("mscIint()\n");
+#endif
+	uint8_t msResult = MS_CBW_PASS;
+
+	CBWTag = 0;
+	uint32_t start = millis();
+	// Check if device is connected.
+	do {
+		if((millis() - start) >= MSC_CONNECT_TIMEOUT) {
+			return MS_NO_MEDIA_ERR;  // Not connected Error.
+		}
+		yield();
+	} while(!available()); 
+	
+	msReset();
+	delay(500);
+	maxLUN = msGetMaxLun();
+//	msResult = msReportLUNs(&maxLUN);
+//Serial.printf("maxLUN = %d\n",maxLUN);
+//	delay(150);
+	//-------------------------------------------------------
+//	msResult = msStartStopUnit(1);
+	msResult = WaitMediaReady();
+	if(msResult)
+		return msResult;
+		
+	// Retrieve drive information.
+	msDriveInfo.initialized = true;
+	msDriveInfo.hubNumber = getHubNumber();			// Which HUB.
+	msDriveInfo.hubPort = getHubPort();				// Which HUB port.
+	msDriveInfo.deviceAddress = getDeviceAddress();	// Device addreess.
+	msDriveInfo.idVendor = getIDVendor();  			// USB Vendor ID.
+	msDriveInfo.idProduct = getIDProduct();  		// USB Product ID.
+	msResult = msDeviceInquiry(&msInquiry);			// Config Info.
+	if(msResult)
+		return msResult;
+	msResult = msReadDeviceCapacity(&msCapacity);	// Size Info.
+	if(msResult)
+		return msResult;
+	memcpy(&msDriveInfo.inquiry, &msInquiry, sizeof(msInquiryResponse_t));
+	memcpy(&msDriveInfo.capacity, &msCapacity, sizeof(msSCSICapacity_t));
+	return msResult;
+}
+
+//---------------------------------------------------------------------------
+// Perform Mass Storage Reset
+void msController::msReset() {
+#ifdef DBGprint
+	Serial.printf("msReset()\n");
+#endif
+	mk_setup(setup, 0x21, 0xff, 0, 0, 0);
```

Comment:
Hi, I am late to the party, just want to say the `wIndex` should  be the interface number. This currently only work with MSC as the first interface (itf num = 0). Same applies to `msGetMaxLun()`. I am not familiar with codebase to submit the PR.